src/jubjub/mod.rs

```diff
@@ -434,6 +460,260 @@ impl JubjubBls12 {
     }
 }
 
+/// Determines whether or not we have to write/overwrite the exp-table file at `exp_table_path`.
+///
+/// If the exp-table file at `exp_table_path` exists and has the correct metadata (number of
+/// segments and window-size) that is desired for these Pedersen Hash params, we return `false` to
+/// indicate that no further exp-table generation needs to take place.
+///
+/// If the file doesn't exist, is not long enough (the number of segments in the exp-table file is
+/// fewer than the number of segments required by the Pedersen Hash params), or the window-size in
+/// the file differs from that being requested, we return `true` to indicate that a new exp-table
+/// must be generated and written.
+fn must_write_exp_table(
+    n_segments_required: usize,
+    window_size_required: u32,
+    exp_table_path: &str,
+) -> bool {
+    let mut file = match File::open(exp_table_path) {
+        Ok(file) => file,
+        _ => return true,
+    };
+
+    let n_segments_in_file = {
+        let mut buf = [0u8; 4];
+        if file.read_exact(&mut buf).is_err() {
+            return true;
+        }
+        u32::from_le_bytes(buf) as usize
+    };
+
+    if n_segments_in_file < n_segments_required {
+        return true;
+    }
+
+    let window_size_in_file = {
+        file.seek(SeekFrom::Start(4)).unwrap();
+        let mut buf = [0u8; 4];
+        if file.read_exact(&mut buf).is_err() {
+            return true;
+        }
+        u32::from_le_bytes(buf)
+    };
+
+    window_size_in_file != window_size_required
+}
+
+/// Create the exp-table for the provided set of generators.
+fn create_exp_table_for_generators(
```

Comment:
it is probably better if we return a `io::Result` here, so that there is some way of handling failed file reads/writes